### PR TITLE
Fixes #19589 - Added plugin migration generator

### PR DESCRIPTION
**Pull request by ShimShtein**
Now plugin creators would be able to generate migrations by invoking

``` sh
rails g plugin:migration <migration name and options> --plugin_name=my_plugin
```

Additional options are available, use `--help` key to display all options.

### CODE REVIEW

File: lib/generators/plugin/migration_generator.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
+module Plugin
+    class MigrationGenerator < Rails::Generators::Base
```
Comment (ohadlevy) [inline lib/generators/plugin/migration_generator.rb]: nitpick: 4 whitespaces instead of 2.

File: lib/generators/plugin/migration_generator.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      def copy_latest_migration
+        destination_file = File.join(
+          destination_root,
+          'db/migrate',
+          destination_file_name
+        )
+
+        unless options['pretend']
+          copy_file migration_file, destination_file
+          remove_file File.join('db/migrate', migration_file)
```
Comment (ohadlevy) [inline lib/generators/plugin/migration_generator.rb]: why not move instead of copy / remove?

File: lib/generators/plugin/migration_generator.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
+module Plugin
+    class MigrationGenerator < Rails::Generators::Base
+      class_option :plugin_name, :required => true
+      class_option :plugin_source
+
+      attr_accessor :migration_pattern
+
+      hook_for :orm, :required => true do |instance, klass|
+        new_migration = instance.invoke klass, nil, nil, destination_root: instance.send(:migration_destination)
+        instance.migration_pattern =  File.basename(new_migration.first).sub(/^\d*_/, '').sub(/\.rb$/, '')
```
Comment (ohadlevy) [inline lib/generators/plugin/migration_generator.rb]: hard to follow, maybe extract to a method?

File: lib/generators/plugin/migration_generator.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      end
+
+      def initialize(*args)
+        super
+
```
Comment (ohadlevy) [inline lib/generators/plugin/migration_generator.rb]: nitpick: extra line

File: lib/generators/plugin/migration_generator.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      def migration_file
+        @migration_file ||= begin
+          full_name = Dir.glob(File.join(migration_destination + 'db/migrate/*.rb')).grep(/\d+_#{migration_pattern}.rb$/).first
+          File.basename(full_name)
+        end
+      end
+
+      def plugin_source
+        @plugin_source ||= options[:plugin_source] || "../#{plugin_name}/"
```
Comment (ohadlevy) [inline lib/generators/plugin/migration_generator.rb]: maybe .. is not great for non development setup's, maybe do it only in development?

**Review by ohadlevy — COMMENTED** [review]
not a full review- just a few comments...

File: lib/generators/plugin/migration_generator.rb (hunk 5 above)
Comment (ShimShtein) [reply lib/generators/plugin/migration_generator.rb]: I see most of usage from developer setup (who else generates migrations?).
In case the folder structure is not standard - you can always specify the path by using `--plugin_source=../where/my/plugin/lives` option from the command line.

**Comment by ShimShtein** [comment]
@ohadlevy, addressed all issues.
Was a bit too eager to get the PR out :)

File: lib/generators/plugin/migration_generator.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      end
+    end
+
+    def plugin_source
+      # assuming a default folder layout for developer setup.
+      @plugin_source ||= options[:plugin_source] || "../#{plugin_name}/"
+    end
+
+    def plugin_name
+      @plugin_name ||= options[:plugin_name].underscore
```
Comment (ares) [inline lib/generators/plugin/migration_generator.rb]: arguments usually use `--plugin-name` notation, the underscore should be replaced with dash. Other migration generator arguments do it at least.

File: lib/generators/plugin/migration_generator.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    def migration_file
+      @migration_file ||= begin
+        full_name = Dir.glob(File.join(migration_destination + 'db/migrate/*.rb')).grep(/\d+_#{migration_pattern}.rb$/).first
+        File.basename(full_name)
+      end
+    end
+
+    def plugin_source
+      # assuming a default folder layout for developer setup.
+      @plugin_source ||= options[:plugin_source] || "../#{plugin_name}/"
```
Comment (ares) [inline lib/generators/plugin/migration_generator.rb]: this should also use dash instead of underscore

File: lib/generators/plugin/migration_generator.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    def destination_file_name
+      File.basename(migration_file, File.extname(migration_file)) +
+        ".#{plugin_name}" +
+        File.extname(migration_file)
+    end
+
+    def migration_file
+      @migration_file ||= begin
+        full_name = Dir.glob(File.join(migration_destination + 'db/migrate/*.rb')).grep(/\d+_#{migration_pattern}.rb$/).first
+        File.basename(full_name)
```
Comment (ares) [inline lib/generators/plugin/migration_generator.rb]: getting ugly trace on this line

```
/home/ares/Projekty/Zdrojaky/foreman/lib/generators/plugin/migration_generator.rb:62:in `basename': no implicit conversion of nil into String (TypeError)
        from /home/ares/Projekty/Zdrojaky/foreman/lib/generators/plugin/migration_generator.rb:62:in `migration_file'
        from /home/ares/Projekty/Zdrojaky/foreman/lib/generators/plugin/migration_generator.rb:51:in `destination_file_name'
        from /home/ares/Projekty/Zdrojaky/foreman/lib/generators/plugin/migration_generator.rb:25:in `copy_latest_migration'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.4/lib/thor/command.rb:27:in `run'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.4/lib/thor/invocation.rb:126:in `invoke_command'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.4/lib/thor/invocation.rb:133:in `block in invoke_all'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.4/lib/thor/invocation.rb:133:in `each'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.4/lib/thor/invocation.rb:133:in `map'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.4/lib/thor/invocation.rb:133:in `invoke_all'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.4/lib/thor/group.rb:232:in `dispatch'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/thor-0.19.4/lib/thor/base.rb:444:in `start'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.8/lib/rails/generators.rb:157:in `invoke'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.8/lib/rails/commands/generate.rb:13:in `<top (required)>'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:274:in `block in require'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:240:in `load_dependency'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:274:in `require'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.8/lib/rails/commands/commands_tasks.rb:123:in `require_command!'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.8/lib/rails/commands/commands_tasks.rb:130:in `generate_or_destroy'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.8/lib/rails/commands/commands_tasks.rb:50:in `generate'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.8/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/railties-4.2.8/lib/rails/commands.rb:17:in `<top (required)>'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/polyglot-0.3.5/lib/polyglot.rb:65:in `require'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:274:in `block in require'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:240:in `load_dependency'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:274:in `require'
        from /home/ares/Projekty/Zdrojaky/foreman/bin/rails:9:in `<top (required)>'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:268:in `load'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:268:in `block in load'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:240:in `load_dependency'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activesupport-4.2.8/lib/active_support/dependencies.rb:268:in `load'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require'
        from /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:54:in `require'
```

the path it constructs is wrong - `/home/ares/Projekty/Zdrojaky/foreman_openscap/../foreman_openscap/db/migrate/*.rb`, note that the foreman dir from which I run it is at `/home/ares/Projekty/Zdrojaky/foreman`. It seems that `migration_destination` returns wrong directory.

**Review by ares — CHANGES_REQUESTED** [review]
It does not seem to be linked to proper ticket. The ticket mentions using scopes for `rake db:migrate`.  Also please reply in the ticket regarding possible duplicity.

File: lib/generators/plugin/migration_generator.rb (hunk 6 above)
Comment (ShimShtein) [reply lib/generators/plugin/migration_generator.rb]: Arguments are converted to dash notation automatically, try it by running `rails g plugin:migration --help`.

For me it returns:
```
@shim (19454)$ rails g plugin:migration --help
Running via Spring preloader in process 22637
Usage:
  rails generate plugin:migration [options]

Options:
      --plugin-name=PLUGIN_NAME        # Indicates when to generate plugin name
      [--plugin-source=PLUGIN_SOURCE]  # Indicates when to generate plugin source
  -o, --orm=NAME                       # Orm to be invoked
                                       # Default: active_record
#snip
```

**Comment by ShimShtein** [comment]
Fixed a small issue with paths, now it should at least show what path it was looking for.
 > note that the foreman dir from which I run it is at /home/ares/Projekty/Zdrojaky/foreman

The file should be created in plugin's directory. Now I set it a bit more verbose, so it should work for custom paths specified by `--plugin-source` option too.

**Comment by ares** [comment]
Please fix the tests

**Comment by ShimShtein** [comment]
Added a test to test arbitrary paths and changed tests to work with `Dir.tmpdir` in order to fix test run failures in Jenkins.

**Comment by ShimShtein** [comment]
Apparently the problem was `stdout` capturing in generators test framework. I have cancelled the capture and the tests seem to pass.

**Comment by ShimShtein** [comment]
Foreman test failure seems unrelated.

[test]

File: lib/generators/plugin/migration_generator.rb (hunk 6 above)
Comment (ares) [reply lib/generators/plugin/migration_generator.rb]: your first example suggested to use `--plugin_name=my_plugin` but you're right, the help is printed correctly :+1:

**Comment by ares** [comment]
Thanks @ShimShtein, I think the PR is ready for merge. It does not solve the issue entirely though so I don't this the commit message should be "Fixes #19454". We still can't use `rake db:migrate SCOPE=my_plugin` until all existing plugin migrations are renamed. I suggest you convert #19454 to "Tracker" and link issue for each plugin (extra points for sending PRs too with renaming) plus the issue for engine engine which you're adding in this PR. Once the commit message is re-linked, I'm happy to merge. After it gets in, I'd like you to document this on plugin writing wiki.

**Review by ares — APPROVED** [review]

**Comment by ShimShtein** [comment]
http://projects.theforeman.org/issues/19454 has been marked as tracker, and this PR moved to http://projects.theforeman.org/issues/19589 as requested.

Test failures are random.

**Comment by ares** [comment]
Thakns @ShimShtein, it looks fine now.

just to be sure

[test foreman]
[test katello]

I haven't seen apache test failures among randomly failing ones

**Comment by ares** [comment]
Katello tests are known issues, foreman tests failed because of npm, rerunning [test foreman]

**Comment by ares** [comment]
Thanks @ShimShtein, merged as fb57b8b09dfb3c968e1e5819e1a332c8eee3ea88, could you please update http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin with instructions how to run the generator and brief explanation why it's important?

**Comment by ShimShtein** [comment]
Done: http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin/diff?utf8=%E2%9C%93&version=116&version_from=115&commit=View+differences

**Comment by ares** [comment]
Thanks!
